wormhole/Dockerfile

```diff
@@ -1,14 +1,22 @@
-FROM pypy:3.9-7.3.9-bullseye
+# Pull official base image from DockerHub
+FROM python:3.9.16-slim-bullseye
 
-COPY Pipfile /app/Pipfile
-COPY Pipfile.lock /app/Pipfile.lock
+# Avoid interactive frontend
+ENV DEBIAN_FRONTEND="noninteractive"
 
-RUN apt update
-RUN apt install rustc -y
+# Install some required packages for testing
+RUN INSTALL_PKGS="git" && \
+    apt-get -q clean && \
+    apt-get -q update && \
+    apt-get install -y $INSTALL_PKGS && \
+    apt-get -q clean
```

Comment:
This is already good, since it
1. runs in a single line / build step (thus resulting in only one Docker "layer") and
2. runs `apt-get clean` so there's less useless stuff in the Docker image.

Further improvements might be
1. Remove the `INSTALL_PKGS` variable since it's only a single package anyway.
2. Run apt-get clean only once, after installing the software.
3. Do you want to also run `apt-get upgrade` or will the `FROM: ` image have all updates always (in my experience they don't).
4. Remove more apt cache files after the clean using `&& rm -rf /var/lib/apt/lists/*` as last step (should save a couple more MB).

Reply:
The image based on `python:3.9.16-slim-bullseye` is now 190MB.
But because @c-chroniko asked for it: the same image based on `python:3.9.16-alpine3.17` is 120MB.

Is it worth it?

Reply:
I think the "new" way to spell all that is just `apt` (instead of `apt-get`). Not sure if this is "new enough" Debian.

I think you _do_ want "apt upgrade" (or "apt dist-upgrade") -- on Stable this should be security fixes etc.

One point of Debian is that everything should be compatible within the version ("11" or "12") so nothing should break by doing upgrades (unless you go from 11 -> 12, for example, but using symbolic names like "stable" instead of codenames like "bullseye" or "bookworm").

there's also "apt autoremove" and "apt autoclean" too if you want to see if that saves a few more MBs.

Reply:
> Yeah! There is even a way to avoid git

Nice! Less code is always better!

> python:3.9.16-alpine3.17

Some people report problems with musl.  I personally like it and tend to use Alpine when I can.

Reply:
> but bump python:3.9.16-slim-bullseye whenever a new version is available/desirable.

Do we need to bump or can we just `FROM:` some `-latest` version like [3.9-bullseye](https://github.com/docker-library/python/blob/331890ef059fae05f84c652520b78c340526dc71/3.9/bullseye/Dockerfile) instead of [3.9.16-bullseye](https://github.com/docker-library/python/blob/331890ef059fae05f84c652520b78c340526dc71/3.9/bullseye/Dockerfile)?

Reply:
> Indeed, apt upgrade will install some security fixes, but will also make the image quite different each time it is built!

Yes it may be somewhat different, but Debian promises that software that works with e.g. "bullseye" should continue to work with "bullseye" as they do bug and security fixes.

So I suppose it depends on the goals here, but for services on the public Internet I would want generally want all the security updates as soon as possible.

Reply:
> > but bump python:3.9.16-slim-bullseye whenever a new version is available/desirable.
> 
> Do we need to bump with each minor release or can we just `FROM:` some `-latest` version like `3.9-slim-bullseye` instead of `3.9.16-slim-bullseye`?

My understanding is:
1. Use `FROM 3.9-slim-bullseye` may give us a different image depending of the time it is built (locally and on the runner). We only need to bump for 3.10...
2. Use `FROM 3.9-16-slim-bullseye` will give us the same image each time (everywhere). We need to bump for 3.10 AND we may want to bump each minor to propagate some updates.

In both cases, someone/something need to commit/merge to re-build and publish the new image!

Reply:
> So I suppose it depends on the goals here, but for services on the public Internet I would want generally want all the security updates as soon as possible.

I guess we could schedule a daily build to automatically create PR whenever a new minor update is available (regardless if we use `3.9` or `3.9.16`).

Reply:
I had a look at the 2 other images (mailbox and relay): there are based on `pypy:3.9-7.3.9-slim-bullseye`.
Thus, same logic there: we do pin a minor (static) version tag rather than a intermediate (dynamic) one.

I'll merge this PR with `FROM 3.9-16-slim-bullseye` for now and create at least one issue to assess if/how to implement a regular update and change (or not) those minor/static tags for all those images.